scalding-core/src/main/scala/com/twitter/scalding/typed/memory_backend/MemoryBackend.scala

```diff
@@ -0,0 +1,527 @@
+package com.twitter.scalding.typed.memory_backend
+
+import cascading.flow.{ FlowDef, FlowConnector }
+import cascading.pipe.Pipe
+import cascading.tap.Tap
+import cascading.tuple.{ Fields, TupleEntryIterator }
+import com.twitter.scalding.typed._
+import com.twitter.scalding.{ Config, Execution, ExecutionCounters, Mode }
+import java.util.concurrent.atomic.AtomicReference
+import java.util.{ ArrayList, Collections, UUID }
+import scala.concurrent.{ Future, ExecutionContext => ConcurrentExecutionContext, Promise }
+import scala.collection.mutable.{ ArrayBuffer, Map => MMap }
+import scala.collection.JavaConverters._
+
+import Execution.{ ToWrite, Writer }
+
+class AtomicBox[T <: AnyRef](init: T) {
+  private[this] val ref = new AtomicReference[T](init)
+
+  def lazySet(t: T): Unit =
+    ref.lazySet(t)
+
+  /**
+   * use a pure function to update the state.
+   * fn may be called more than once
+   */
+  def update[R](fn: T => (T, R)): R = {
+
+    @annotation.tailrec
+    def loop(): R = {
+      val init = ref.get
+      val (next, res) = fn(init)
+      if (ref.compareAndSet(init, next)) res
```

Comment:
yes. But I don't think that can happen because we only mutate it when jobs finish and there are only a finite number of jobs running at a time, so I think it is exponentially unlikely that they do spin forever.